File: cpp/test/src/unit/test_searcher.cpp
// test_searcher.cpp — Unit tests for deglib::search::Searcher (searcher.h)

#include "deglib/builder.h"
#include "deglib/distances.h"
#include "deglib/graph.h"
#include "deglib/optimization.h"
#include "deglib/search/searcher.h"
#include "gtest/gtest.h"

#include <algorithm>
#include <cstdint>
#include <numeric>
#include <random>
#include <span>
#include <vector>

TEST(SearcherTest, DirectFP32Search) {
    const uint32_t dim = 4;
    const uint32_t count = 20;

    std::vector<float> data(count * dim);
    for (size_t i = 0; i < data.size(); ++i) {
        data[i] = static_cast<float>(i * 1.5f + 0.1f);
    }

    auto feature_space = deglib::distances::FloatSpace(dim, deglib::distances::Metric::FP32_L2);
    auto graph = deglib::builder::build_from_data(std::span<const float>(data), dim, {}, 8, deglib::distances::Metric::FP32_L2);

    auto searcher = deglib::search::make_searcher(graph.internal());

    // 1. Raw buffer search
    std::vector<uint32_t> out_indices(5);
    std::vector<float> out_distances(5);
    uint32_t found = searcher->search_f32(data.data(), 5, 0.1f, 1.0f, out_indices.data(), out_distances.data());

    EXPECT_EQ(found, 5u);
    EXPECT_EQ(out_indices[0], 0u);
    EXPECT_NEAR(out_distances[0], 0.0f, 1e-5f);
    for (size_t i = 1; i < 5; ++i) {
        EXPECT_GE(out_distances[i], out_distances[i - 1]);
    }

    // 2. Modern C++20 span / SearchResult search
    auto span_res = searcher->search(std::span<const float>(data.data(), dim), 5, 0.1f, 1.0f, /*return_distances=*/true);
    EXPECT_EQ(span_res.size(), 5u);
    EXPECT_EQ(span_res.indices[0], 0u);
    EXPECT_NEAR(span_res.distances[0], 0.0f, 1e-5f);
}

TEST(SearcherTest, QuantizedInt8WithFP16Refiner) {
    const uint32_t dim = 8;
    const uint32_t count = 50;

    std::mt19937 rng(42);
    std::uniform_real_distribution<float> dist(-1.0f, 1.0f);

    std::vector<float> data(count * dim);
    for (auto& v : data) v = dist(rng);

    auto base_graph = deglib::builder::build_from_data(std::span<const float>(data), dim, {}, 8, deglib::distances::Metric::FP32_L2);

    // Quantize to INT8
    auto quantizer = deglib::optimization::make_scalar_quantizer_int8(data.data(), count, dim);
    std::vector<int8_t> int8_data(count * dim);
    quantizer.quantize(data.data(), int8_data.data(), count, dim);

    auto target_space = deglib::distances::FloatSpace(dim, deglib::distances::Metric::Int8_L2);
    auto ro_graph = deglib::graph::convert_to_readonly_graph(base_graph.internal(), target_space, int8_data.data());

    // Refiner with FP16 base features
    std::vector<uint16_t> base_fp16(count * dim);
    deglib::distances::fp16::floats_to_fp16(data.data(), base_fp16.data(), data.size());
    auto rerank_space = deglib::distances::FloatSpace(dim, deglib::distances::Metric::FP16_L2);

    using RefinerT = deglib::search::ExactRefiner<uint16_t>;

    auto searcher = deglib::search::make_searcher(
        ro_graph,
        quantizer,
        RefinerT(rerank_space, base_fp16.data(), count)
    );

    // Query for 5th item
    auto q = std::span<const float>(data.data() + 5 * dim, dim);
    auto res = searcher->search(q, 5, /*eps=*/0.2f, /*rerank_factor=*/2.0f, /*return_distances=*/true);

    EXPECT_EQ(res.size(), 5u);
    EXPECT_EQ(res.indices[0], 5u);
    EXPECT_NEAR(res.distances[0], 0.0f, 1e-3f);
}

TEST(SearcherTest, QuantizedUint8WithFP32Refiner) {
    const uint32_t dim = 8;
    const uint32_t count = 50;

    std::mt19937 rng(42);
    std::uniform_real_distribution<float> dist(0.0f, 10.0f);

    std::vector<float> data(count * dim);
    for (auto& v : data) v = dist(rng);

    auto base_graph = deglib::builder::build_from_data(std::span<const float>(data), dim, {}, 8, deglib::distances::Metric::FP32_L2);

    auto quantizer = deglib::optimization::make_scalar_quantizer_uint8(data.data(), count, dim);
    std::vector<uint8_t> uint8_data(count * dim);
    quantizer.quantize(data.data(), uint8_data.data(), count, dim);

    auto target_space = deglib::distances::FloatSpace(dim, deglib::distances::Metric::Uint8_L2);
    auto ro_graph = deglib::graph::convert_to_readonly_graph(base_graph.internal(), target_space, uint8_data.data());

    auto rerank_space = deglib::distances::FloatSpace(dim, deglib::distances::Metric::FP32_L2);

    using RefinerT = deglib::search::ExactRefiner<float>;

    auto searcher = deglib::search::make_searcher(
        ro_graph,
        quantizer,
        RefinerT(rerank_space, data.data(), count)
    );

    auto q = std::span<const float>(data.data() + 2 * dim, dim);
    auto res = searcher->search(q, 3, /*eps=*/0.2f, /*rerank_factor=*/1.5f, /*return_distances=*/true);

    EXPECT_GE(res.size(), 1u);
    EXPECT_EQ(res.indices[0], 2u);
    EXPECT_NEAR(res.distances[0], 0.0f, 1e-4f);
}

TEST(SearcherTest, EVPQuantizerWithFP32Refiner) {
    const uint32_t dim = 16;
    const uint32_t count = 50;
    const uint32_t non_zeros = 4;

    std::mt19937 rng(42);
    std::normal_distribution<float> dist(0.0f, 1.0f);

    std::vector<float> data(count * dim);
    for (size_t i = 0; i < count; ++i) {
        float norm_sq = 0.0f;
        for (size_t d = 0; d < dim; ++d) {
            data[i * dim + d] = dist(rng);
            norm_sq += data[i * dim + d] * data[i * dim + d];
        }
        float inv_norm = 1.0f / std::sqrt(norm_sq);
        for (size_t d = 0; d < dim; ++d) {
            data[i * dim + d] *= inv_norm;
        }
    }

    auto base_graph = deglib::builder::build_from_data(std::span<const float>(data), dim, {}, 8, deglib::distances::Metric::FP32_InnerProduct);

    deglib::quantization::evp::EvpQuantizer quantizer(non_zeros);
    auto evp_data = quantizer.quantize(data.data(), count, dim);
    auto target_space = deglib::distances::FloatSpace(dim, deglib::distances::Metric::EVP_InnerProduct);
    auto ro_graph = deglib::graph::convert_to_readonly_graph(base_graph.internal(), target_space, evp_data.data());

    auto rerank_space = deglib::distances::FloatSpace(dim, deglib::distances::Metric::FP32_InnerProduct);

    using QuantT = deglib::quantization::evp::EvpQuantizer;
    using RefinerT = deglib::search::ExactRefiner<float>;

    auto searcher = deglib::search::make_searcher(
        ro_graph,
        quantizer,
        RefinerT(rerank_space, data.data(), count)
    );

    auto q = std::span<const float>(data.data() + 3 * dim, dim);
    auto res = searcher->search(q, 3, /*eps=*/0.3f, /*rerank_factor=*/2.0f, /*return_distances=*/true);

    EXPECT_GE(res.size(), 1u);
    EXPECT_EQ(res.indices[0], 3u);
    EXPECT_NEAR(res.distances[0], 0.0f, 1e-4f);

    // FP16 query test with EVPQuantizer
    std::vector<uint16_t> q_fp16(dim);
    deglib::distances::fp16::floats_to_fp16(data.data() + 3 * dim, q_fp16.data(), dim);
    auto res_fp16 = searcher->search(std::span<const uint16_t>(q_fp16.data(), dim), 3, /*eps=*/0.3f, /*rerank_factor=*/2.0f, /*return_distances=*/true);
    EXPECT_GE(res_fp16.size(), 1u);
    EXPECT_EQ(res_fp16.indices[0], 3u);
    EXPECT_NEAR(res_fp16.distances[0], 0.0f, 1e-3f);
}

TEST(SearcherTest, FlatBatchAndIntoSearch) {
    const uint32_t dim = 4;
    const uint32_t count = 30;

    std::vector<float> data(count * dim);
    for (size_t i = 0; i < data.size(); ++i) {
        data[i] = static_cast<float>(i * 0.5f + 0.1f);
    }

    auto graph = deglib::builder::build_from_data(std::span<const float>(data), dim, {}, 8, deglib::distances::Metric::FP32_L2);
    auto searcher = deglib::search::make_searcher(graph.internal());

    const size_t n_queries = 5;
    const uint32_t k = 4;

    // 1. search_batch returning SearchResultBatch (flat buffer)
    auto flat_res = searcher->search_batch(
        std::span<const float>(data.data(), n_queries * dim), n_queries, k,
        /*eps=*/0.1f, /*rerank_factor=*/1.0f,
        /*threads=*/2, /*return_distances=*/true
    );
    EXPECT_EQ(flat_res.size(), n_queries);
    EXPECT_EQ(flat_res.k, k);
    EXPECT_EQ(flat_res.indices.size(), n_queries * k);
    EXPECT_EQ(flat_res.distances.size(), n_queries * k);

    for (size_t q = 0; q < n_queries; ++q) {
        auto q_indices = flat_res.get_indices(q);
        auto q_dists = flat_res.get_distances(q);
        EXPECT_EQ(q_indices[0], static_cast<uint32_t>(q));
        EXPECT_NEAR(q_dists[0], 0.0f, 1e-5f);
    }

    // 2. search_batch into preallocated buffers
    std::vector<uint32_t> preallocated_indices(n_queries * k, 9999);
    std::vector<float> preallocated_dists(n_queries * k, 9999.0f);
    searcher->search_batch(
        std::span<const float>(data.data(), n_queries * dim),
        n_queries, k,
        std::span<uint32_t>(preallocated_indices),
        std::span<float>(preallocated_dists),
        /*eps=*/0.1f, /*rerank_factor=*/1.0f,
        /*threads=*/1,
        /*return_distances=*/true
    );

    for (size_t q = 0; q < n_queries; ++q) {
        EXPECT_EQ(preallocated_indices[q * k], static_cast<uint32_t>(q));
        EXPECT_NEAR(preallocated_dists[q * k], 0.0f, 1e-5f);
    }
}

File: python/tests/test_searcher.py
import numpy as np
import pytest
import deglib
from deglib import Metric, FloatSpace, create_searcher


def test_searcher_direct_f32():
    np.random.seed(42)
    dim = 16
    n = 100
    data = np.random.randn(n, dim).astype(np.float32)

    graph = deglib.builder.build_from_data(data, edges_per_vertex=16, metric=Metric.FP32_L2)
    searcher = create_searcher(graph)

    query = data[0]
    res = searcher.search(query, k=5, eps=0.1)
    assert len(res) == 5
    assert res[0] == 0

    # With return_distances
    indices, distances = searcher.search(query, k=5, eps=0.1, return_distances=True)
    assert len(indices) == 5
    assert len(distances) == 5
    assert indices[0] == 0
    assert distances[0] == pytest.approx(0.0, abs=1e-5)
    # Check ascending order
    for i in range(1, len(distances)):
        assert distances[i] >= distances[i - 1]

    # With unsorted
    indices_unsorted, distances_unsorted = searcher.search(query, k=5, eps=0.1, return_distances=True, unsorted=True)
    assert len(indices_unsorted) == 5
    assert set(indices_unsorted) == set(indices)

    # Unified search method with 1D, (1, dim), and (dim, 1) queries
    res_1xDim = searcher.search(query.reshape(1, -1), k=5, eps=0.1)
    assert res_1xDim.shape == (5,)
    assert res_1xDim[0] == 0

    res_Dimx1 = searcher.search(query.reshape(-1, 1), k=5, eps=0.1)
    assert res_Dimx1.shape == (5,)
    assert res_Dimx1[0] == 0

    # Unified search method with 2D batch queries
    res_batch_unified = searcher.search(data[:10], k=5, eps=0.1, threads=2)
    assert res_batch_unified.shape == (10, 5)
    assert res_batch_unified[0, 0] == 0

    indices_batch, dists_batch = searcher.search(data[:10], k=5, eps=0.1, threads=2, return_distances=True)
    assert indices_batch.shape == (10, 5)
    assert dists_batch.shape == (10, 5)
    assert indices_batch[0, 0] == 0
    assert dists_batch[0, 0] == pytest.approx(0.0, abs=1e-5)


def test_searcher_quantized_int8_with_rerank():
    np.random.seed(42)
    dim = 32
    n = 200
    data = np.random.randn(n, dim).astype(np.float32)

    # 1. Base graph in FP32
    graph = deglib.builder.build_from_data(data, edges_per_vertex=16, metric=Metric.FP32_L2)

    # 2. INT8 Quantization
    quantizer = deglib.optimization.make_scalar_quantizer_int8(data)
    int8_data = quantizer.quantize(data)
    target_space = FloatSpace.create(dim=dim, metric=Metric.Int8_L2)
    ro_graph = graph.to_readonly(target_space, int8_data)

    # 3. Refine features in FP16
    base_fp16 = deglib.distances.floats_to_fp16(data)
    rerank_space = FloatSpace.create(dim=dim, metric=Metric.FP16_L2)

    # 4. Create Searcher
    searcher = create_searcher(
        graph=ro_graph,
        quantizer=quantizer,
        refine_space=rerank_space,
        refine_data=base_fp16,
    )

    query = data[5]
    res = searcher.search(query, k=5, eps=0.2, rerank_factor=1.5)
    assert len(res) == 5
    assert res[0] == 5

    # With return_distances
    indices, distances = searcher.search(query, k=5, eps=0.2, rerank_factor=1.5, return_distances=True)
    assert indices[0] == 5
    assert distances[0] == pytest.approx(0.0, abs=1e-3)
    for i in range(1, len(distances)):
        assert distances[i] >= distances[i - 1]

    # Batch search
    res_batch = searcher.search(data[:5], k=5, eps=0.2, rerank_factor=1.5, threads=2)
    assert res_batch.shape == (5, 5)
    for i in range(5):
        assert res_batch[i, 0] == i


def test_searcher_quantized_uint8_with_rerank():
    np.random.seed(42)
    dim = 32
    n = 200
    data = np.random.uniform(0.0, 10.0, size=(n, dim)).astype(np.float32)

    graph = deglib.builder.build_from_data(data, edges_per_vertex=16, metric=Metric.FP32_L2)

    quantizer = deglib.optimization.make_scalar_quantizer_uint8(data)
    uint8_data = quantizer.quantize(data)
    target_space = FloatSpace.create(dim=dim, metric=Metric.Uint8_L2)
    ro_graph = graph.to_readonly(target_space, uint8_data)

    searcher = create_searcher(
        graph=ro_graph,
        quantizer=quantizer,
        refine_space=FloatSpace.create(dim=dim, metric=Metric.FP32_L2),
        refine_data=data,
    )

    query = data[12]
    res = searcher.search(query, k=5, eps=0.2, rerank_factor=1.5)
    assert len(res) == 5
    assert res[0] == 12

    indices, distances = searcher.search(query, k=5, eps=0.2, rerank_factor=1.5, return_distances=True)
    assert indices[0] == 12
    assert distances[0] == pytest.approx(0.0, abs=1e-5)


def test_searcher_evp_quantizer():
    np.random.seed(42)
    dim = 64
    n = 200
    data = np.random.randn(n, dim).astype(np.float32)
    # L2 normalize for MIPS / InnerProduct
    data = data / np.linalg.norm(data, axis=1, keepdims=True)

    graph = deglib.builder.build_from_data(data, edges_per_vertex=16, metric=Metric.FP32_InnerProduct)

    # 16 non-zeros for EVP
    non_zeros = 16
    evp_data = deglib.optimization.quantize_batch(data, non_zeros=non_zeros)
    target_space = FloatSpace.create(dim=dim, metric=Metric.EVP_InnerProduct)
    ro_graph = graph.to_readonly(target_space, evp_data)

    searcher = create_searcher(
        graph=ro_graph,
        quantizer=non_zeros,
        refine_space=FloatSpace.create(dim=dim, metric=Metric.FP32_InnerProduct),
        refine_data=data,
    )

    query = data[3]
    res = searcher.search(query, k=3, eps=0.3, rerank_factor=2.0)
    assert len(res) == 3
    assert res[0] == 3

    # FP16 query
    query_fp16 = deglib.distances.floats_to_fp16(query)
    res_fp16, dists_fp16 = searcher.search(query_fp16, k=3, eps=0.3, rerank_factor=2.0, return_distances=True)
    assert len(res_fp16) == 3
    assert res_fp16[0] == 3
    assert dists_fp16[0] == pytest.approx(0.0, abs=1e-3)
